Refactor Blocks

mandaiy opened the pull request:
Following changes are made in this commit:
- Rename classe
  As we discussed offline
- Extract common functions among block classes
  Changed NumberTextHolder interface for better encapsulation
- Change return value type of BlockBase#getKind
  The previous code uses Class class, which does not represent actual
  use case (e.g., ForwardSecBlock#getKind returns SequenceBlock.class).
  Replaced Class<? extends Block> type with Enum.
- Remove locale-dependent code on parsing numbers
  The previous code relies on the Locale.US no matter what locale
  this app runs on.
  This commit fixes the locale dependent code by extracting the parsing
  logic.
- Fix the unspoken rule that using scaled values
  for communicating between blocks and number pickers.
    The previous code scaled values so that the value has no radix
    point, which was confusing code.
    Fixed by replacing the type with BigDecimal and explicitly use the raw
    value.
    In accordance with it, changed the type of "number"
    in ProgramDataSpec for saving without data loss
- Remove code in Unit class for handling quantities, which was ugly and cannot be used widely.
  Changed second string ("second/s" -> just "s")

mandaiy committed:
Refactor Blocks

Following changes are made in this commit:
* Rename classes
    As we discussed offline

* Extract common functions among block classes
    Changed NumberTextHolder interface for better encapsulation

* Change return value type of BlockBase#getKind
    The previous code uses Class class, which does not represent actual
    use case (e.g., ForwardSecBlock#getKind returns SequenceBlock.class).
    Replaced Class<? extends Block> type with Enum.

* Remove locale-dependent code on parsing numbers
    The previous code relies on the Locale.US no matter what locale
    this app runs on.
    This commit fixes the locale dependent code by extracting the parsing
    logic.

* Fix the unspoken rule that using scaled values
  for communicating between blocks and number pickers.
    The previous code scaled values so that the value has no radix
    point, which was confusing code.
    Fixed by replacing the type with BigDecimal and explicitly use the raw
    value.
    In accordance with it, changed the type of "number"
    in ProgramDataSpec for saving without data loss

* Remove code in Unit class for handling quantities, which was ugly and cannot be used widely.
    Changed second string ("second/s" -> just "s")

tiwanari commented:
Thanks for this PR 👍  Because I merged other PRs, there are conflicts :( Please fix them.

mandaiy commented:
Hey @tiwanari, resolved conflicts.
Please take a look at my commit. Thanks.

tiwanari commented:
Thanks for notifying me :) Please give me a little time to check them.

tiwanari commented on app/src/main/java/com/pileproject/drive/programming/visual/block/BlockFactory.java:
```diff
@@ -38,131 +38,93 @@
+
         try {
             return (Class<T>) Class.forName(className);
+
```
I don't think this blank line makes this program more readable. Is there any specific reason to keep blank lines like this? If no, please remove such lines.

mandaiy replied:
Seriously? I DO THINK SO.

but this is kinda religious war and here's your realm.

tiwanari replied:
I'm sorry, I didn't want to mean it :( I agree that adequate blank lines help code reading.

I couldn't see the consistency between these blank lines here and previous codes at review but now I'm guessing you added blank lines between long lines. It can give a kind of tender impression to codes I think. I wanted to know some of rules, i.e. why you added it here and didn't there.

We haven't talked enough about blank lines beforehand and formatters never specify it. I really appreciate if you share your thoughts :)

mandaiy replied:
Don't take my comment seriously and offensive, please.
For consistency you are right.
And topics of the format are bit off from here.
Anyway I'll merge this. Thanks.

incidentally you can witch-hunt with

```
$ grep -ElR "catch \(.*\) {" app | xargs perl -0777 -pe 's/\n(\s*} catch \(.*\) {)\n/$1/g'
```

for making your realm calm and getting rid of "pagans"

tiwanari replied:
Thanks. The one-liner seems to illustrate a rule "add blank lines before & after a catch clause" :)

btw, `realm' has a bit overwhelmed impression for me... :sweat:

tiwanari commented on app/src/main/java/com/pileproject/drive/programming/visual/block/BlockFactory.java:
```diff
@@ -38,131 +38,93 @@
     public static final int UNDO = 3;
     public static final int LOAD = 4;
 
-    /**
```
This is just a question. Did you remove Javadocs from here because the method is `private`, right?

mandaiy replied:
The answer is Yes

tiwanari commented on app/src/main/java/com/pileproject/drive/programming/visual/block/repetition/RepetitionBlock.java:
```diff
@@ -29,14 +30,14 @@
 public abstract class RepetitionBlock extends BlockBase {
 
-    public static final int FOREVER_WHILE_OFFSET = -1000;
+    static final int FOREVER_WHILE_OFFSET = -1000;
```
I found this cannot be `package private` here because it is used in ExecuteCodition. This may be controversial.

mandaiy replied:
Moved to ExecutionCondition, since this field is irrelevant for the operation of blocks.
(btw ExecutionCondition is no longer "condition")

tiwanari commented on app/src/main/java/com/pileproject/drive/programming/visual/block/BlockFactory.java:
```diff
@@ -38,131 +38,93 @@
     public static final int SEQUENCE = 0;
     public static final int REPETITION = 1;
     public static final int SELECTION = 2;
+    @Deprecated
```
Sorry, this is my fault. I forgot to remove this constant. Please remove it and I will add it in #12 .

mandaiy replied:
Removed

tiwanari commented on app/src/main/java/com/pileproject/drive/programming/visual/block/NumberTextViewDelegate.java:
```diff
@@ -0,0 +1,75 @@
+import java.math.BigDecimal;
+import java.util.Locale;
+
+public class NumberTextViewDelegate {
```
Could you give a short description of this class as a Javadoc here?

mandaiy replied:
Added

tiwanari commented on app/src/main/java/com/pileproject/drive/programming/visual/layout/BlockSpaceManagerBase.java:
```diff
@@ -126,7 +127,7 @@ public void loadUserProgram(String programName) {
-    public abstract void addBlocks(ArrayList<BlockBase> blocks);
+    public abstract void addBlocks(List<BlockBase> blocks);
 
     private void placeBlocks(ArrayList<BlockBase> data) {
```
According to the signature change of `addBlocks`, this should also be changed (from ArrayList to List) to keep consistency.

mandaiy replied:
Changed

tiwanari commented on app/src/main/java/com/pileproject/drive/programming/visual/layout/ProgrammingSpaceManager.java:
```diff
@@ -115,85 +141,35 @@ private void moveViewWithinItsParent(View view, int currentX, int currentY) {
-            numText.setOnLongClickListener(new OnTouchNumTextListener(block));
-        }
-    }
+    private class OnTouchNumTextListener implements View.OnLongClickListener {
```
Please fix the name of this class from *NumText* to *NumberText*.

mandaiy replied:
Changed

tiwanari commented on app/src/main/java/com/pileproject/drive/programming/visual/layout/ProgrammingSpaceManager.java:
```diff
@@ -115,85 +141,35 @@ private void moveViewWithinItsParent(View view, int currentX, int currentY) {
-        AlphaAnimation alpha = new AlphaAnimation(1, 0);
-        alpha.setDuration(1000);
-        alpha.setInterpolator(new CycleInterpolator(3));
+        private final Context context;
```
Here, this class has private members and they should be named with a name starting with `m` like `mContext`.

tiwanari replied:
This is controversial especially for inner classes.

mandaiy replied:
Renamed

tiwanari commented on app/src/main/java/com/pileproject/drive/util/development/Unit.java:
```diff
@@ -16,66 +16,46 @@
 
-            case Percentage:
-                return value + " " + context.getResources().getString(R.string.percent);
+    public String getUnitString(BigDecimal value, int precision) {
```
This method name is somewhat confusing for me (I guessed this method returns just a unit in String like "%").

mandaiy replied:
Renamed method name to "decorateValue"

tiwanari commented on app/src/main/java/com/pileproject/drive/util/string/ParseUtil.java:
```diff
@@ -0,0 +1,78 @@
+import java.text.ParseException;
+import java.util.Locale;
+
+public class ParseUtil {
```
This class also has many blank lines. It's okay if it has reasons.

tiwanari commented on app/src/main/java/com/pileproject/drive/util/string/ParseUtil.java:
```diff
@@ -0,0 +1,78 @@
+        }
+    }
+
+    public static BigDecimal bigDecimalValueOf(String value) throws NumberFormatException {
```
Please give a Javadocs as you made above.

mandaiy replied:
Added

tiwanari commented on app/src/main/java/com/pileproject/drive/view/NumberSelectWheelScrollerView.java:
```diff
@@ -34,83 +35,140 @@
-     * @param context      an application context
-     * @param integralPart the length of integral part
-     * @param decimalPart  the length of decimal part
+     * * Constructor.
```
nit: double *.

mandaiy replied:
Removed

tiwanari commented on app/src/nxt/java/com/pileproject/drive/programming/visual/block/sequence/SetLeftMotorSpeedBlock.java:
```diff
@@ -17,59 +17,47 @@
-        return Integer.parseInt(numText.getText().toString());
-    }
+    // TODO: set from preference
+    private static final int PRECISION = 1;
```
1 is okay but it was 0.

mandaiy replied:
Changed

tiwanari commented on app/src/nxt/java/com/pileproject/drive/programming/visual/block/sequence/TurnLeftSecBlock.java:
```diff
@@ -17,64 +17,49 @@
     }
 
     @Override
     public int action(
```
I know this is my fault but if you have time, please remove break lines... (this is not mandatory)

mandaiy replied:
Fixed

tiwanari reviewed (changes requested):
I made comments but before seeing them, please make the PR enable to be built. There are errors with merged code (ExecutionCondition).

tiwanari commented:
Overall, they are really good and cleaned 👍

mandaiy committed:
Fix PR according to the comments on #35

* remove UNDO in BlockFactory.java
* move FOREVER_WHILE_OFFSET in RepetitionBlock.java to ExecutionCondition.java
* add Javadocs on NumberTextDelegate
* change the signature of BlockSpaceManagerBase#placeBlock (ArrayList to List)
* rename OnTouchNumTextListener to OnTouchNumberTextListener
* remove OnNumberTextListener#context and rename holder to mHolder
* rename Unit#getUnitString to decorateValue and introduce new class NumberUtil

mandaiy committed:
Add and fix tests

for the following classes:
* NumberTextViewDelegate
* Unit
* NumberUtil
* ParseUtil

And cradle.build in the root directory is modified to add the dexmaker libraries
which are necessary for mockito

mandaiy committed:
Remove superfluous line-breaks

mandaiy committed:
Remove an extra comma

mandaiy committed:
Add constructor guards for util classes

tiwanari reviewed (approved):
LGTM!
